doc/platform/nlapi/netlink_api_kernel_programming.md

```diff
@@ -0,0 +1,223 @@
+  
+# Table of Contents  
+* [List of Tables](#list-of-tables)  
+* [Revision](#revision)  
+* [About This Manual](#about-this-manual)  
+* [Scope](#scope)  
+* [Definition/Abbreviation](#definitionabbreviation)   
+* [Requirement Overview](#requirement-overview)  
+* [Functionality](#functionality)  
+* [Design](#design)  
+* [Flow Diagrams](#flow-diagrams)   
+* [Serviceability and Debug](#serviceability-and-debug)  
+* [Warm Boot Support](#warm-boot-support)  
+* [Scalability](#scalability)  
+* [Unit Test](#unit-test)
+  
+# List of Tables  
+[Table 1: Abbreviations](#table-1-abbreviations)  
+[Table 2: Performance comparison](#table-2-performance-comparison)  
+[Table 3: Unit test cases](#table-3-unit-test-cases)  
+
+  
+# Revision  
+| Rev | Date | Author | Change Description |  
+|:---:|:-----------:|:------------------:|-----------------------------------|  
+| 0.1 | 09/24/2019 | Naveen Suvarna | Initial version |  
+  
+# About this Manual  
+This document provides general information about the SONiC Linux kernel programming performance enhancement using Netlink based APIs.  
+
+# Scope  
+This document describes the high level design of  Linux kernel programming performance enhancement using netlink based API calls replacing system() calls executing iproute2 commands on linux shell. 
+
+# Definition/Abbreviation  
+### Table 1: Abbreviations  
+| **Term** | **Meaning** |  
+|--------------------------|-------------------------------------|  
+| NL | Netlink |  
+| API | Application Program Interface |
+| NLAPI | Netlink API |
+
+
+
+  
+# 1 Requirement Overview  
+## 1.1 Functional Requirements  
+1. Support Netlink API calls to replace all exec calls of iproute2 ip and bridge commands found in SONiC modules. 
+ 
+## 1.2 Configuration and Management Requirements  
+NA
+  
+## 1.3 Scalability Requirements  
+NA
+  
+## 1.4 Warm Boot Requirements  
+NA  
+  
+# 2 Functionality  
+## 2.1 Target Deployment Use Cases  
+Performance improvement in scaled configuration during bulk config or system restart.
+  
+## 2.2 Functional Description  
+This document provides functional design and specifications of kernel programming performance enhancement by replacing system() calls with netlink APIs for kernel interaction.
+ 
+SONiC modules program the kernel constructs like bridge, VLAN, IP-link, VRF, IP Address into kernel using unix iproute-2 utility commands. Within a sonic process, this is being done by forking new process using fork (thru popen) system call and then executing required  iproute2 ip and/or bridge commands in it. This system() call invocation is slow (compared to netlink APIs) and executed synchronously and process executing this waits for its completion, impacting the performance. Kernel programming using this method is primarily used in config managers of SwSS components. Config managers predominantly use the popen call based kernel programming during user configuration or config replay at restart or system bringup. 
```

Comment:
Does this include LAG, LAG member as well?teammgr in sonic-swss creates LAG and LAG members using 'ip link'